### PR TITLE
Reduce FeatureReference checking, lazy init FormFeature

### LINKED ISSUE #677: FormFeature allocates when unused
form-feature-allocation

Resolved by: https://github.com/aspnet/HttpAbstractions/pull/675

### LINKED ISSUE #880: Don't allocate the FormFeature eagerly per request
Today, in order to pass the form options to the `FormFeature`, we eagerly allocate it here https://github.com/aspnet/HttpAbstractions/blob/163836fe1f31dd1d2b288659c2dcaf1d827da3c4/src/Microsoft.AspNetCore.Http/HttpContextFactory.cs#L49. We should find a creative way to avoid doing this so eagerly in the common cases. Here are some options:

- Invent a new way to pass things to features lazily (hard, we tried this and ended up with crazy things)
- Only new up the `FormOptions` if they were overridden. We'd need a way to detect if the `FormOptions` were actually set. This is doable but could get ugly as there's no way to detect if options were configured. /cc @HaoK 
- Only set the `FormFeature` if the request is a POST and `HasFormContentType`. Gross, but would be more pay for play than what we have today.

/cc @vancem @Tratcher

### LINKED ISSUE #924: FeatureCache is over-checked and over-reset
A simple plantext middleware as below, will check the `Frame` `IFeatureCollection` 7 times and initialize/reset the Context, Resquest and Responses feature caches twice each.
```csharp
public static Task WriteResponse(HttpResponse response)
{
    var payloadLength = _helloWorldPayload.Length;
    response.StatusCode = 200;
    response.ContentType = "text/plain";
    response.ContentLength = payloadLength;
    return response.Body.WriteAsync(_helloWorldPayload, 0, payloadLength);
}
```

The extra reset and checks are from initalization; then reinitialization after `RequestServicesFeature` is added to the collection by the host, increasing the `IFeatureCollection`'s version count.

### PR DESCRIPTION
**Pull request by benaadams**
Update to https://github.com/aspnet/HttpAbstractions/pull/925

Resolves #924 "FeatureCache is over-checked and over-reset"
Resolves #677 "FormFeature allocates when unused"
Resolves #880 "Don't allocate the FormFeature eagerly per request"

/cc @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNetCore.Http/DefaultHttpContext.cs [hunk 1]
```diff
@@ -44,15 +46,26 @@ public DefaultHttpContext()
         }
 
         public DefaultHttpContext(IFeatureCollection features)
+            : this(new FeatureCollection(), FormOptions.Default)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Http/DefaultHttpContext.cs]: Shouldn’t we be passing the feature collection and not a new one?

File: test/Microsoft.AspNetCore.Http.Tests/DefaultHttpContextTests.cs [hunk 2]
```diff
@@ -158,7 +158,9 @@ public void UpdateFeatures_ClearsCachedFeatures()
             features.Set<IHttpWebSocketFeature>(new TestHttpWebSocketFeature());
 
             // featurecollection is set. all cached interfaces are null.
-            var context = new DefaultHttpContext(features);
+            var context = new DefaultHttpContext(features, FormOptions.Default);
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Http.Tests/DefaultHttpContextTests.cs]: Make sure we validate the default ctor as well

**Review by davidfowl — CHANGES_REQUESTED** [review]

**Comment by benaadams** [comment]
Need to unpick this a bit more; lot of calls going on 

(prior)
image

File: test/Microsoft.AspNetCore.Http.Tests/DefaultHttpContextTests.cs (hunk 2 above)
Comment (benaadams) [reply test/Microsoft.AspNetCore.Http.Tests/DefaultHttpContextTests.cs]: Done

File: src/Microsoft.AspNetCore.Http/DefaultHttpContext.cs (hunk 1 above)
Comment (benaadams) [reply src/Microsoft.AspNetCore.Http/DefaultHttpContext.cs]: Done

**Comment by benaadams** [comment]
Before
image

After
image

Still a bucket load of calls; but less

File: src/Microsoft.AspNetCore.Http/Features/FormOptions.cs [hunk 3]
```diff
@@ -8,6 +8,8 @@ namespace Microsoft.AspNetCore.Http.Features
 {
     public class FormOptions
     {
+        public static readonly FormOptions Default = new FormOptions();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http/Features/FormOptions.cs]: We can't make a mutable instance a public static

Comment (benaadams) [reply src/Microsoft.AspNetCore.Http/Features/FormOptions.cs]: Good point

File: src/Microsoft.AspNetCore.Http/Internal/DefaultHttpRequest.cs [hunk 4]
```diff
@@ -15,26 +15,39 @@ public class DefaultHttpRequest : HttpRequest
         // Lambdas hoisted to static readonly fields to improve inlining https://github.com/dotnet/roslyn/issues/13624
         private readonly static Func<IFeatureCollection, IHttpRequestFeature> _nullRequestFeature = f => null;
         private readonly static Func<IFeatureCollection, IQueryFeature> _newQueryFeature = f => new QueryFeature(f);
-        private readonly static Func<HttpRequest, IFormFeature> _newFormFeature = r => new FormFeature(r);
+        private readonly static Func<(HttpRequest request, FormOptions options), IFormFeature> _newFormFeature = (r) => new FormFeature(r.request, r.options ?? FormOptions.Default);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Http/Internal/DefaultHttpRequest.cs]: Tuples? 😄

Comment (benaadams) [reply src/Microsoft.AspNetCore.Http/Internal/DefaultHttpRequest.cs]: Custom struct or another Fetch overload?

**Comment by muratg** [comment]
@davidfowl @Tratcher is this still relevant/useful in your opinion? If so, let's move it forward. If not, let's get it closed.

Ben, thanks for the PR!

**Comment by davidfowl** [comment]
@benaadams I'm going to move this to 3.0

**Review by davidfowl — APPROVED** [review]

**Comment by natemcmaster** [comment]
@davidfowl I'm about to merge this repo into aspnet/AspNetCore. If you'd like this PR merged, can you bring it in today? If not, I'm going to close this and you'll have to re-open the PR.

**Comment by davidfowl** [comment]
@natemcmaster close it and I'll make @benaadams reopen it 😄

**Pull request closed without merging**